src/md_lammps.c

```diff
@@ -473,4 +474,527 @@ static md_molecule_loader_i lammps_api = {
+		md_file_seek(traj_data->file, beg, MD_FILE_BEG);
+		const int64_t bytes_read = md_file_read(traj_data->file, &ptr[1], frame_size);
+		md_mutex_unlock(&traj_data->mutex);
+		ASSERT(frame_size == bytes_read);
+	}
+
+	return frame_size;
+}
+
+bool lammps_decode_frame_data(struct md_trajectory_o* inst, const void* frame_data_ptr, int64_t frame_data_size, md_trajectory_frame_header_t* header, float* x, float* y, float* z) {
+	ASSERT(inst);
+	ASSERT(frame_data_ptr);
+	ASSERT(frame_data_size);
+
+	str_t tokens[32];
+	int64_t num_tokens = 0;
+	int64_t num_token_types = 0;
+
+	const int64_t* ptr = frame_data_ptr;
+	int32_t frame_idx = (int32_t)ptr[0];
+
+
+	md_lammps_trajectory_t* traj_data = (md_lammps_trajectory_t*)inst;
+	if (traj_data->magic != MD_LAMMPS_TRAJ_MAGIC) {
+		MD_LOG_ERROR("Error when decoding frame header, lammps magic did not match");
+		return false;
+	}
+	if (frame_idx < 0 || frame_idx >= traj_data->header.num_frames) {
+		MD_LOG_ERROR("Error when decoding frame data, corrupt frame index");
+		return false;
+	}
+
+	str_t str = { .ptr = (const char*)(ptr + 1), .len = frame_data_size };
+	str_t line;
+
+	//Parse the ITEM: ATOMS to see what type if coordinates we are working with
+	str_t coord_type = { 0 };
+	int32_t coord_start = -1;
+	if (str_extract_line(&line, &str)) {
+		if (str_equal_cstr_n(line, "ITEM: ATOMS", 11)) {
+			num_tokens = extract_tokens(tokens, 32, &line);
+			num_token_types = num_tokens - 2;
+			if (num_token_types < 3) {
+				MD_LOG_ERROR("Fewer than 3 token types in ITEM: ATOMS line");
+				return false;
+			}
+			for (int32_t i = 0; i < num_tokens; i++) {
+				//Use the first complete coord combo. x,y,z,xs,ys picks x,y,z and x,y,xs,ys,zs picks xs,ys,zs
+				if (
+					str_equal_cstr(tokens[i], "x") && str_equal_cstr(tokens[i + 1], "y") && str_equal_cstr(tokens[i + 2], "z") ||
+					str_equal_cstr(tokens[i], "xs") && str_equal_cstr(tokens[i + 1], "ys") && str_equal_cstr(tokens[i + 2], "zs") ||
+					str_equal_cstr(tokens[i], "xu") && str_equal_cstr(tokens[i + 1], "yu") && str_equal_cstr(tokens[i + 2], "zu") ||
+					str_equal_cstr(tokens[i], "xs") && str_equal_cstr(tokens[i + 1], "ys") && str_equal_cstr(tokens[i + 2], "zs")) {
+					coord_type = tokens[i];
+					coord_start = i - 2; //Skip ITEM: ATOMS tokens
+					break;
+				}
+			}
+			if (coord_start == -1) {
+				MD_LOG_ERROR("Could not parse coord definition in ITEM: ATOMS line");
+				return false;
+			}
+			//step = (int32_t)parse_int(str_substr(line, 10, 4));
+		}
+	}
+
+	//Extract coordinates and scale them if the type requires it. We leave unwrapped coordinates as is
+	int64_t i = 0;
+	while (str_extract_line(&line, &str) && i < traj_data->header.num_atoms) {
+		num_tokens = extract_tokens(tokens, 32, &line);
+		if (coord_start + 2 + 1 > num_tokens) {
+			MD_LOG_ERROR("Not enough tokens in atom line");
+			return false;
+		}
+		if (str_equal_cstr(coord_type, "x") || str_equal_cstr(coord_type, "xu")) { //Unscaled coordinates
```

Comment:
parsing of trajectories is more performance-critical than parsing topologies, so I would rather see that we extract some flags ITEM: ATOMS. if we should scale or not or something. The branches are predictable in this case, but they can be avoided.

Reply:
Implemented a coord_type enum that we parse once from the file, together with a coord start. This is saved in cache and only read once. Then we perform a batch scaling if needed on the read x,y,z arrays